Feat/1st stage training script

deep-diver opened the pull request:
base code of training script for 1st stage SFT model

sayakpaul commented on sft_1st_stage.py:
```diff
@@ -0,0 +1,125 @@
+        "WANDB_PROJECT" in os.environ and len(os.environ["WANDB_PROJECT"]) > 0
+    )
+    # Only overwrite environ if wandb param passed
+    if len(args.wandb_project) > 0:
```
I think it's better to default this CLI argument to None and condition on that. When am argument is provided then we log everything relevant. WDYT?

sayakpaul commented on sft_1st_stage.py:
```diff
@@ -0,0 +1,125 @@
+    parser.add_argument("--lora_r", type=int, default=6)
+    parser.add_argument("--lora_alpha", type=int, default=32)
+    parser.add_argument("--lora_dropout", type=float, default=0.05)
+    parser.add_argument("--lora_target_modules", nargs="+", default=["c_proj", "c_attn", "q_attn"])
```
Using the linear modules for LoRA might worth trying too.

deep-diver replied:
I see there are couple of linear layers. which one do you have in mind?

I also noticed `q_attn` does not exist. my bad

```
GPTBigCodeForCausalLM(
  (transformer): GPTBigCodeModel(
    (wte): Embedding(49152, 6144)
    (wpe): Embedding(8192, 6144)
    (drop): Dropout(p=0.1, inplace=False)
    (h): ModuleList(
      (0-39): 40 x GPTBigCodeBlock(
        (ln_1): LayerNorm((6144,), eps=1e-05, elementwise_affine=True)
        (attn): GPTBigCodeAttention(
          (c_attn): Linear(in_features=6144, out_features=6400, bias=True)
          (c_proj): Linear(in_features=6144, out_features=6144, bias=True)
          (attn_dropout): Dropout(p=0.1, inplace=False)
          (resid_dropout): Dropout(p=0.1, inplace=False)
        )
        (ln_2): LayerNorm((6144,), eps=1e-05, elementwise_affine=True)
        (mlp): GPTBigCodeMLP(
          (c_fc): Linear(in_features=6144, out_features=24576, bias=True)
          (c_proj): Linear(in_features=24576, out_features=6144, bias=True)
          (act): GELUActivation()
          (dropout): Dropout(p=0.1, inplace=False)
        )
      )
    )
    (ln_f): LayerNorm((6144,), eps=1e-05, elementwise_affine=True)
  )
  (lm_head): Linear(in_features=6144, out_features=49152, bias=False)
)
```

sayakpaul replied:
Pretty cool thread here: https://twitter.com/mrm8488/status/1672355317487640577?s=20

deep-diver replied:
so, in this architecture, `c_attn, `c_proj`, `c_fc`, and `lm_head` could be applied

sayakpaul commented on sft_1st_stage.py:
```diff
@@ -0,0 +1,125 @@
+    )
+    trainer.train()
+    
+    trainer.model.save_pretrained(os.path.join(args.output_dir, "final_checkpoint/"))
```
Let's also have a utility for pushing to the HF hub. WDYT?

sayakpaul reviewed (approved):
LGTM with a few nits.

deep-diver commented:
addressed your comments @sayakpaul 

about pushing the model to hub, I added relevant arguments to `TrainingArgument`, `push_to_hub`, `hub_model_id`, and `hub_token`.

about the comment below,
> Using the linear modules for LoRA might worth trying too.

You mean you want to include them by default?

sayakpaul commented on sft_1st_stage.py:
```diff
@@ -0,0 +1,139 @@
+    # Only overwrite environ if wandb param passed
+    if args.wandb_project:
+        os.environ["WANDB_PROJECT"] = args.wandb_project
+    if len(args.wandb_watch) > 0:
+        os.environ["WANDB_WATCH"] = args.wandb_watch
+    if args.wandb_log_model:
+        os.environ["WANDB_LOG_MODEL"] = str(args.wandb_log_model)
```
Shouldn't this be conditioned on `if args.wandb_project:`? Also, I would be more explicit - `if args.wandb_project is not None:` (nit).

sayakpaul commented on sft_1st_stage.py:
```diff
@@ -0,0 +1,139 @@
+        os.environ["WANDB_LOG_MODEL"] = str(args.wandb_log_model)
+    
+    if args.push_to_hub:
+        if not hub_model_id:
+            print("You set `push_to_hub=True`, but `hub_model_id` is not provided.")
```
For the `hub_model_id` we could do something like:

https://github.com/huggingface/diffusers/blob/61916fefc400d1b77171490b6ff7d53e509ed9de/examples/dreambooth/train_dreambooth_lora.py#L745

Then we could remove the `hub_token` argument easily and just instruct the user that they should run `huggingface-cli login` if `push_to_hub` is True.

WDYT?

deep-diver replied:
Could you elaborate? The reference you shared seems like using custom `create_repo` function, and it still requires `token` argument.

sayakpaul reviewed (commented):
Some more comments :)

I can add utilities for pushing the adapter weights to the Hub and for running some validation inference during training to track progress as we do here: https://github.com/huggingface/diffusers/blob/61916fefc400d1b77171490b6ff7d53e509ed9de/examples/dreambooth/train_dreambooth_lora.py#L1253.

deep-diver commented:
@sayakpaul 
> I can add utilities for pushing the adapter weights to the Hub and for running some validation inference during training to track progress as we do here

that would be great! let's do that after merging this PR!

sayakpaul reviewed (approved):
Thanks a lot for iterating.
